### PR TITLE
Feature/time extractor

### PR DESCRIPTION
**Pull request by StokesMIDE**
This reorganizes `idelib.util` a little and adds `idelib.util.extractTime()`. A couple of minor tweaks elsewhere.

Since the Lab and the `endaq-python-ide` package need `extractTime()`, it made sense to move it to the package they share: `idelib`. It's a fairly general-purpose utility function.

### CODE REVIEW

File: idelib/util.py [hunk 1]
```diff
@@ -35,80 +37,154 @@ def verify(data, schema=None):
...
+
+                if writePrev:
+                    # Write the previous block, which is before the extraction interval.
+                    # This is to ensure that initial data is not left out.
+                    prev = lastBlocks.get(chId, None)
+                    if prev:
+                        data = prev[0].getRaw()
+                        fs.write(data)
+                        copiedBytes += len(data)
+                        channelsWritten[chId] += 1
```
Comment (CFlaniganMide) [inline idelib/util.py:165]: I'm not sure I understand why we want to do this.  I would expect to have no data on a channel if it has no data within the specified range, rather than some data before the cutoff.

File: idelib/util.py [hunk 2]
```diff
@@ -35,80 +37,154 @@ def verify(data, schema=None):
...
+        pass
+
+    finally:
+        if isinstance(out, (str, Path)):
+            fs.close()
+
+    if updater:
+        updater(done=True)
+
+    return copiedBytes, sum(channelsWritten.values())
```
Comment (CFlaniganMide) [inline idelib/util.py:189]: Would it be preferable to ask the file stream how many bytes have been written rather than keep track?

```suggestion
    return fs.tell(), sum(channelsWritten.values())
```

File: idelib/util.py [hunk 3]
```diff
@@ -35,80 +37,154 @@ def verify(data, schema=None):
...
-            v, elementType = d[1]
-        else:
-            k, v = d
-            for t, elementType in attTypes:
-                if isinstance(v, t):
+    if isinstance(doc, (str, Path)):
+        doc = openFile(doc)
+
+    if endTime is None:
+        endTime = sys.maxsize
```
Comment (CFlaniganMide) [inline idelib/util.py:74]: Since python supports arbitrarily large integers, and `sys.maxsize + 1` returns a valid integer, would this make more sense?

```suggestion
        endTime = float('infinity')
```

**Review by CFlaniganMide — CHANGES_REQUESTED** [review]

File: idelib/util.py (hunk 1 above)
Comment (StokesMIDE) [reply idelib/util.py:165]: Primarily, this is for the temperature-compensating bivariate polynomials. Also, I believe users will expect some data, even if the channel doesn't have samples that happen to fall within the specified interval. There have been some complaints when they 'lose' data from the start or end of a split recording.

File: idelib/util.py (hunk 2 above)
Comment (StokesMIDE) [reply idelib/util.py:189]: Since it's already written, I don't see much of an advantage in changing it.... also, what if the stream it was passed for output didn't start at `fs.tell() == 0`, or it is a file-like object without `tell()`?

File: idelib/util.py (hunk 1 above)
Comment (CFlaniganMide) [reply idelib/util.py:165]: The bivariates definitely make sense.  It looks like if all the blocks on a channel end before the start of the interval you're extracting, it could end up writing none anyway.

File: idelib/util.py (hunk 2 above)
Comment (CFlaniganMide) [reply idelib/util.py:189]: Hadn't thought that it might be a stream-like that had already been written to, that's a good point.

File: idelib/util.py (hunk 1 above)
Comment (CFlaniganMide) [reply idelib/util.py:165]: I still think that this could cause an issue if someone wants to export acceleration data that relies on the internal temperature, but only exists after the temperature sensor has no data.

**Review by CFlaniganMide — CHANGES_REQUESTED** [review]

File: idelib/util.py (hunk 1 above)
Comment (StokesMIDE) [reply idelib/util.py:165]: Unless the file is damaged, there will always be temperature data before the analog accelerometer channel that needs it. Even if the last temperature sample came before the extracted range, the file will contain that last block. If the method `EventArray.getEventIndexNear()` (used by the bivariate polynomial) is called with a time beyond the actual range, it will return either the first or last.

tl;dr: It shouldn't be an issue, and if it is, it's a bug elsewhere in `idelib`.

**Review by CFlaniganMide — APPROVED** [review]
